applications: Added guide about extending Matter bridge app (#13452)
nrfconnect/sdk-nrf: applications/matter_bridge/doc/adding_ble_bridged_device_service.rst

               gatt_desc = bt_gatt_dm_desc_by_uuid(discoveredData, gatt_chrc, sUuidCcc);
               if (!gatt_desc) {
                  return -EINVAL;
               }
               mCccHandle = gatt_desc->handle;

               VerifyOrReturn(mDevice.mConn, LOG_ERR("Invalid connection object"));

               /* Configure subscription for the button characteristic */
               mGattSubscribeParams.ccc_handle = mCccHandle;
               mGattSubscribeParams.value_handle = mCharacteristicHandle;
               mGattSubscribeParams.value = BT_GATT_CCC_NOTIFY;
               mGattSubscribeParams.notify = MyBtServiceDataProvider::GattNotifyCallback;
               mGattSubscribeParams.subscribe = nullptr;
               mGattSubscribeParams.write = nullptr;
               return bt_gatt_subscribe(mDevice.mConn, &mGattSubscribeParams);
            }

#. Include ``MyBtServiceDataProvider`` implementation created in a previous steps to the compilation.
   To do that, edit :file:`CMakeLists.txt` file, as follows:

   .. code-block:: cmake

      target_sources(app PRIVATE
        src/ble_providers/my_bt_service_data_provider.cpp
      )

> #. Provide allocator for ``MyBtServiceDataProvider`` object creation.

Reviewer: ```suggestion
#. Provide an allocator for ``MyBtServiceDataProvider`` object creation.
```

>    The Matter Bridge application uses a :c:struct:`BleBridgedDeviceFactory` factory module that creates a pair of ``Matter Bridged Device`` and ``Bridged Device Data Provider`` objects matching specific Matter device type id.

Reviewer: ```suggestion
   The Matter Bridge application uses a :c:struct:`BleBridgedDeviceFactory` factory module that creates a pair of ``Matter Bridged Device`` and ``Bridged Device Data Provider`` objects matching a specific Matter device type ID.
```

>    To add support for creating the ``MyBtServiceDataProvider`` object, edit :file:`src/ble_providers/ble_bridged_device_factory.h` and :file:`src/ble_providers/ble_bridged_device_factory.cpp` files as follows:

Reviewer: ```suggestion
   To add support for creating the ``MyBtServiceDataProvider`` object, edit the :file:`src/ble_providers/ble_bridged_device_factory.h` and :file:`src/ble_providers/ble_bridged_device_factory.cpp` files as follows:
```

   - :file:`ble_bridged_device_factory.h`

      .. code-block:: C++

         #include "my_bt_service_data_provider.h"

   - :file:`ble_bridged_device_factory.cpp`, :c:func:`GetDataProviderFactory`

      .. code-block:: C++

		   { ServiceUuid::MyBtService, [](UpdateAttributeCallback updateClb, InvokeCommandCallback commandClb) {
			   return chip::Platform::New<MyBtServiceDataProvider>(updateClb, commandClb);
		   } },

#. Provide mapping between the ``My Bt Service`` UUID and corresponding Matter device types in the helper methods.

>    a. Add ``MyBtService`` UUID in the :c:enum:`ServiceUuid` declaration, in :file:`src/ble_providers/ble_bridged_device_factory.h`.

Reviewer: ```suggestion
   a. Add the ``MyBtService`` UUID in the :c:enum:`ServiceUuid` declaration in the :file:`src/ble_providers/ble_bridged_device_factory.h` header file.
```

>    #. Perform proper mapping of Bluetooth UUID and Matter device types in :c:func:`MatterDeviceTypeToBleService` and :c:func:`BleServiceToMatterDeviceType` methods, in :file:`src/ble_providers/ble_bridged_device_factory.cpp`.

Reviewer: ```suggestion
   #. Perform proper mapping of Bluetooth UUID and Matter device types in the :c:func:`MatterDeviceTypeToBleService` and :c:func:`BleServiceToMatterDeviceType` methods in the :file:`src/ble_providers/ble_bridged_device_factory.cpp` file.
```

.. _matter_bridge_app_extending_ble_service:

Adding support for a new Bluetooth LE service
#############################################

The Matter Bridge application supports bridging Bluetooth LE devices using :ref:`LED Button Service <lbs_readme>` and :ref:`Environmental Sensing Service <peripheral_esp>` Bluetooth services.
You can also add support for a proprietary Bluetooth LE service, if required by your use case.
The functionality of the added Bluetooth LE service has to be represented by one or more device types available in the :ref:`Matter Data Model <ug_matter_device_types>`.
For example, the :ref:`LED Button Service <lbs_readme>` is represented by the Matter On/Off Light and Matter Generic Switch device types.

You can bridge the new Bluetooth LE service with one or more Matter device types supported by the Matter Bridge application, or add support for new Matter device type.

> To learn how to add support for new Matter device type in the Matter Bridge application, see the :ref:`matter_adding_bridged_matter_device` page.

Author: Sure, I haven't had time to verify the build process yesterday. I will do that in the meantime of your reviews.

               if (!gatt_chrc) {
                  return -EINVAL;
               }

               gatt_desc = bt_gatt_dm_desc_by_uuid(discoveredData, gatt_chrc, sUuidChar);
               if (!gatt_desc) {
                  return -EINVAL;
               }
               mCharacteristicHandle = gatt_desc->handle;

               gatt_desc = bt_gatt_dm_desc_by_uuid(discoveredData, gatt_chrc, sUuidCcc);
               if (!gatt_desc) {
                  return -EINVAL;
               }
               mCccHandle = gatt_desc->handle;

               VerifyOrReturn(mDevice.mConn, LOG_ERR("Invalid connection object"));

               /* Configure subscription for the button characteristic */
               mGattSubscribeParams.ccc_handle = mCccHandle;
               mGattSubscribeParams.value_handle = mCharacteristicHandle;
               mGattSubscribeParams.value = BT_GATT_CCC_NOTIFY;
               mGattSubscribeParams.notify = MyBtServiceDataProvider::GattNotifyCallback;
               mGattSubscribeParams.subscribe = nullptr;
               mGattSubscribeParams.write = nullptr;
               return bt_gatt_subscribe(mDevice.mConn, &mGattSubscribeParams);
            }

> #. Include the ``MyBtServiceDataProvider`` implementation created in a previous steps to the compilation.

Reviewer 2: Ditto. Add instead of include?

      Learn the constructor signature, that will be used in the child class implemented in the next steps.
   #. Add a new :c:struct:`MyBtServiceDataProvider` class inheriting :c:struct:`BLEBridgedDeviceProvider`, and implement its constructor in the :file:`my_bt_service_data_provider.cpp` and :file:`my_bt_service_data_provider.h` files.

      .. code-block:: C++

        #include "ble_bridged_device.h"
        #include "ble_connectivity_manager.h"
        #include "bridged_device_data_provider.h"

        #include "my_bt_service.h"

        class MyBtServiceDataProvider : public BLEBridgedDeviceProvider {
        public:
            explicit MyBtServiceDataProvider(UpdateAttributeCallback updateCallback, InvokeCommandCallback commandCallback) : BLEBridgedDeviceProvider(updateCallback, commandCallback) {}

        };

   #. Open the :file:`nrf/samples/matter/common/bridge/ble_bridged_device.h` header file again to see which methods of :c:struct:`BLEBridgedDeviceProvider` class are purely virtual (assigned with ``=0``) and have to be overridden by the :c:struct:`MyBtServiceDataProvider` class.

>       Note that :c:struct:`BLEBridgedDeviceProvider` inherits from the :c:struct:`BridgedDeviceDataProvider` class, so the :c:struct:`MyBtServiceDataProvider` class has to implement the purely virtual methods of :c:struct:`BridgedDeviceDataProvider` as well.

Reviewer: ```suggestion

      Note that :c:struct:`BLEBridgedDeviceProvider` inherits from the :c:struct:`BridgedDeviceDataProvider` class, so the :c:struct:`MyBtServiceDataProvider` class has to implement the purely virtual methods of :c:struct:`BridgedDeviceDataProvider` as well.
```

>    #. Edit the :c:struct:`MyBtServiceDataProvider` class, in the :file:`my_bt_service_data_provider.h` header file to declare the required methods as follows:

Reviewer: ```suggestion
   #. Edit the :c:struct:`MyBtServiceDataProvider` class in the :file:`my_bt_service_data_provider.h` header file to declare the required methods as follows:
```

      .. code-block:: C++

        void Init() override;
        void NotifyUpdateState(chip::ClusterId clusterId, chip::AttributeId attributeId, void *data,
                    size_t dataSize) override;
        CHIP_ERROR UpdateState(chip::ClusterId clusterId, chip::AttributeId attributeId, uint8_t *buffer) override;
        bt_uuid *GetServiceUuid() override;
        int ParseDiscoveredData(bt_gatt_dm *discoveredData) override;

   #. Implement the body of the :c:func:`Init` method so that it can prepare the data provider for further operation.
      If there are no additional actions to be done before starting the provider, it can be implemented in the :file:`my_bt_service_data_provider.cpp` file as empty.

      .. code-block:: C++

        void MyBtServiceDataProvider::Init()
        {
            /* Do nothing in this case */
        }

   #. Implement the body of the :c:func:`NotifyUpdateState` method that shall be called after every data change related to the Matter devices bridged to the Bluetooth LE device using ``My Bt Service``.
      It is used to inform the ``Bridge Manager`` and Matter Data Model that an attribute value should be updated.

>       To make the method invoke appropriate callback, edit :file:`my_bt_service_data_provider.cpp` file as follows:

Reviewer: ```suggestion
      To make the method invoke the appropriate callback, edit the :file:`my_bt_service_data_provider.cpp` file as follows:
```

      .. code-block:: C++

        void MyBtServiceDataProvider::NotifyUpdateState(chip::ClusterId clusterId, chip::AttributeId attributeId, void *data,
                            size_t dataSize)
        {
            if (mUpdateAttributeCallback) {
                mUpdateAttributeCallback(*this, clusterId, attributeId, data, dataSize);
            }
        }

   #. Implement the body of the :c:func:`UpdateState` method.

>       This will be called by the ``Bridge Manager`` to inform that data in Matter Data Model was changed and request propagating this information to the Bluetooth LE end device.

Reviewer: ```suggestion
      This will be called by the ``Bridge Manager`` to inform that data in Matter Data Model was changed and request propagating this information to the Bluetooth LE end device.

```

      The content of this method depends on the supported Matter device types and the Bluetooth characteristics supported by the specific Bluetooth LE profile.
      If the profile supports write operations, the implementation should analyze the Matter :c:var:`clusterId` and :c:var:`attributeId` variables, and perform a Bluetooth GATT write operation to the corresponding Bluetooth characteristic.
      Otherwise, the method can be left empty.

>       To handle write operations to the Bluetooth LE device, edit :file:`my_bt_service_data_provider.h` and :file:`my_bt_service_data_provider.cpp` files using the following code snippets:

Reviewer: ```suggestion
      To handle write operations to the Bluetooth LE device, edit the :file:`my_bt_service_data_provider.h` and :file:`my_bt_service_data_provider.cpp` files using the following code snippets:
```

                  return;
               }

               /* Save data received in GATT write response. */
               /* ... */
               DeviceLayer::PlatformMgr().ScheduleWork(NotifyAttributeChange, reinterpret_cast<intptr_t>(provider));
            }

            void MyBtServiceDataProvider::NotifyAttributeChange(intptr_t context)
            {
               MyBtServiceDataProvider *provider = reinterpret_cast<MyBtServiceDataProvider *>(context);
               /* Invoke provider->NotifyUpdateState() method to inform the `Bridge Manager` that write operation suceeded and Matter Data Model state can be updated. */
               /* ... */
            }

   #. Implement the body of the :c:func:`GetServiceUuid` method.
      This shall return the UUID of the ``My Bt Service`` Bluetooth LE service.
      To do this, edit the :file:`my_bt_service_data_provider.cpp` file as follows:

      .. code-block:: C++

         static bt_uuid *sServiceUuid = BT_UUID_MY_BT_SERVICE;
         bt_uuid *MyBtServiceDataProvider::GetServiceUuid()
         {
            return sServiceUuid;
         }

   #. Implement the body of the :c:func:`ParseDiscoveredData` method.

>       This should parse the input data and save required Bluetooth characteristic handles for further use.

Reviewer: ```suggestion
      This should parse the input data and save the required Bluetooth characteristic handles for further use.
```

      The Bluetooth LE service can support different sets of characteristics, so the method content will depend on this set.
      Additionally, the Bluetooth LE service might support subscriptions through the GATT CCC characteristic.
      In that case, the method implementation should establish a subscription session with the Bluetooth LE end device.

>       For example, to handle a single characteristic that additionally supports subscriptions, edit the :file:`my_bt_service_data_provider.h` and :file:`my_bt_service_data_provider.cpp` files as follows:

Reviewer: ```suggestion

      For example, to handle a single characteristic that additionally supports subscriptions, edit the :file:`my_bt_service_data_provider.h` and :file:`my_bt_service_data_provider.cpp` files as follows:
```

               if (!gatt_desc) {
                  return -EINVAL;
               }
               mCharacteristicHandle = gatt_desc->handle;

               gatt_desc = bt_gatt_dm_desc_by_uuid(discoveredData, gatt_chrc, sUuidCcc);
               if (!gatt_desc) {
                  return -EINVAL;
               }
               mCccHandle = gatt_desc->handle;

               VerifyOrReturn(mDevice.mConn, LOG_ERR("Invalid connection object"));

               /* Configure subscription for the button characteristic */
               mGattSubscribeParams.ccc_handle = mCccHandle;
               mGattSubscribeParams.value_handle = mCharacteristicHandle;
               mGattSubscribeParams.value = BT_GATT_CCC_NOTIFY;
               mGattSubscribeParams.notify = MyBtServiceDataProvider::GattNotifyCallback;
               mGattSubscribeParams.subscribe = nullptr;
               mGattSubscribeParams.write = nullptr;
               return bt_gatt_subscribe(mDevice.mConn, &mGattSubscribeParams);
            }

#. Add the ``MyBtServiceDataProvider`` implementation created in a previous steps to the compilation process.

>    To do that, edit the :file:`CMakeLists.txt` file, as follows:

Reviewer: ```suggestion
   To do that, edit the :file:`CMakeLists.txt` file as follows:
```

   .. code-block:: cmake

      target_sources(app PRIVATE
        src/ble_providers/my_bt_service_data_provider.cpp
      )

#. Provide an allocator for ``MyBtServiceDataProvider`` object creation.

>    The Matter Bridge application uses a :c:struct:`BleBridgedDeviceFactory` factory module that creates a pair of ``Matter Bridged Device`` and ``Bridged Device Data Provider`` objects matching a specific Matter device type ID.

Reviewer: ```suggestion
   The Matter Bridge application uses a :c:struct:`BleBridgedDeviceFactory` factory module that creates paired ``Matter Bridged Device`` and ``Bridged Device Data Provider`` objects matching a specific Matter device type ID.

```

.. _matter_adding_ble_bridged_device_service:

Adding support for bridging of new Bluetooth LE service
*******************************************************

> The Matter Bridge application supports bridging Bluetooth LE devices using :ref:`LED Button Service <lbs_readme>` and :ref:`Environmental Sensing Service <peripheral_esp>` Bluetooth services.

Reviewer: I agree we don't need to repeat information. But I do think we could add a link target to the definition of a bridge that we have on the network topology page, and link the word "bridge device" on the main application page and the word "bridging" in this introduction to that definition.

1. Include the header file containing Bluetooth LE service declaration of :c:struct:`bt_uuid` type in the :file:`app_task.cpp` file.

   .. code-block:: C++

      #include <my_bt_service.h>

#. Update the :c:var:`sUuidServices` array in the :file:`app_task.cpp` file to include the Bluetooth LE service UUID.

   .. code-block:: C++

      static bt_uuid *sUuidMyBtService = BT_UUID_MY_BT_SERVICE;
      static bt_uuid *sUuidServices[] = { sUuidLbs, sUuidEs, sUuidMyBtService};

#. Implement the ``Bridged Device Data Provider`` role.

   a. Create :file:`my_bt_service_data_provider.cpp` and :file:`my_bt_service_data_provider.h` files for your Bluetooth LE Data Provider in the :file:`src/ble_providers` directory.
   #. Open the :file:`nrf/samples/matter/common/bridge/ble_bridged_device.h` to see :c:struct:`BLEBridgedDeviceProvider` class constructor.
   #. Add new :c:struct:`MyBtServiceDataProvider` class inherting from the :c:struct:`BLEBridgedDeviceProvider` and implement its constructor in the :file:`my_bt_service_data_provider.cpp` and :file:`my_bt_service_data_provider.h` files.

>       .. code-block:: C++

Reviewer 3: Hmm, ok, I checked it twice and indeed it is under the same indent.
